File: backend/core/phone.py
"""
U.S. phone number utilities — standard (XXX) XXX-XXXX.

Canonical storage: **10 digits, no formatting**. We deliberately keep
this US-only. International numbers / extensions / SMS delivery formats
live elsewhere and are not touched by this module.

`format_us_phone` is *permissive* on read: it only re-formats when the
input normalises to exactly 10 digits. Everything else (legacy seeds
like `+1-555-0102`, blank values, non-US numbers) echoes back
unchanged so old data renders safely without corruption.
"""
from __future__ import annotations

import re

_DIGITS_RE = re.compile(r"\D+")


def _digits_only(value: str | None) -> str:
    if value is None:
        return ""
    return _DIGITS_RE.sub("", str(value))


def normalize_us_phone(value: str | None) -> str | None:
    """Return the canonical 10-digit string, or None if the input is
    empty / blank. Raises `ValueError` for non-empty inputs that don't
    normalise cleanly to 10 digits.

    Accepts: `6155551212`, `615-555-1212`, `(615) 555-1212`,
    `615.555.1212`, `+1 615 555 1212`, `1-615-555-1212` (leading US
    country code stripped).
    """
    if value is None:
        return None
    raw = str(value).strip()
    if raw == "":
        return None
    digits = _digits_only(raw)
    # Tolerate a leading US "+1"/"1" country code.
    if len(digits) == 11 and digits.startswith("1"):
        digits = digits[1:]
    if len(digits) != 10:
        raise ValueError(
            "Phone number must be a 10-digit U.S. number.",
        )
    return digits


def is_valid_us_phone(value: str | None) -> bool:
    """True iff `value` is empty OR normalises to exactly 10 digits."""
    if value is None or str(value).strip() == "":
        return True
    try:
        normalize_us_phone(value)
    except ValueError:
        return False
    return True


def format_us_phone(value: str | None) -> str:
    """Format for display.

    * Exactly 10 US digits (after normalisation)  → `(XXX) XXX-XXXX`.
    * Everything else (None, empty, non-US, 7-digit, extensions,
      garbage) → the input echoed back **unchanged** as str(value).

    This permissiveness is intentional: we render legacy/seed values
    (`+1-555-0102`) without mangling them even though they don't match
    our new canonical shape.
    """
    if value is None:
        return ""
    raw = str(value)
    if raw.strip() == "":
        return ""
    try:
        digits = normalize_us_phone(raw)
    except ValueError:
        return raw
    if digits is None:
        return raw
    return f"({digits[0:3]}) {digits[3:6]}-{digits[6:10]}"


def search_normalize_phone(value: str | None) -> str:
    """Strip formatting for search — returns just the digits the user
    typed, without enforcing length. An empty string means "no phone
    search"; the caller is responsible for deciding the minimum length
    that counts as an intentional query."""
    return _digits_only(value)

File: frontend/src/utils/phone.js
/**
 * U.S. phone number utilities — canonical form `(XXX) XXX-XXXX`.
 *
 * Mirrors `backend/core/phone.py`. US-only by design; international
 * numbers and extension handling live elsewhere. Display is
 * permissive — only 10-digit values get reformatted, everything else
 * echoes back unchanged so legacy data never gets mangled.
 */

const DIGITS_ONLY_RE = /\D+/g;

function digitsOnly(value) {
  if (value == null) return "";
  return String(value).replace(DIGITS_ONLY_RE, "");
}

/**
 * Return the canonical 10-digit string (or null for empty input).
 * Returns null when the input is non-empty but doesn't look like a
 * US phone — call sites can treat that as "still typing / invalid"
 * without throwing.
 */
export function normalizePhone(value) {
  if (value == null) return null;
  const raw = String(value).trim();
  if (raw === "") return null;
  let digits = digitsOnly(raw);
  if (digits.length === 11 && digits[0] === "1") digits = digits.slice(1);
  return digits.length === 10 ? digits : null;
}

/**
 * True iff the value is empty OR normalises to exactly 10 digits.
 * Matches the backend's `is_valid_us_phone` semantics.
 */
export function isValidPhone(value) {
  if (value == null) return true;
  const raw = String(value).trim();
  if (raw === "") return true;
  return normalizePhone(raw) !== null;
}

/**
 * Format a stored phone value for display. 10-digit canonical → pretty
 * `(XXX) XXX-XXXX`. Anything else echoes the original value unchanged,
 * so legacy formats (`+1-555-0102`) render safely.
 */
export function formatPhoneDisplay(value) {
  if (value == null) return "";
  const raw = String(value);
  if (raw.trim() === "") return "";
  const digits = normalizePhone(raw);
  if (!digits) return raw;
  return `(${digits.slice(0, 3)}) ${digits.slice(3, 6)}-${digits.slice(6, 10)}`;
}

/**
 * Strip formatting for search APIs. No length enforcement here — the
 * caller decides what counts as a real query.
 */
export function searchNormalize(value) {
  return digitsOnly(value);
}

/**
 * Pretty-format an in-progress value as the user types: `(XXX)`,
 * `(XXX) XXX`, `(XXX) XXX-XXXX`. Always displays at most 10 digits of
 * input. Designed to be wired into `onChange` via:
 *
 *   const handleChange = (e) => setPhone(formatAsTyped(e.target.value));
 *
 * Storage should still use `normalizePhone` before submitting.
 */
export function formatAsTyped(value) {
  const digits = digitsOnly(value).slice(0, 10);
  if (digits.length === 0) return "";
  if (digits.length <= 3) return `(${digits}`;
  if (digits.length <= 6) return `(${digits.slice(0, 3)}) ${digits.slice(3)}`;
  return `(${digits.slice(0, 3)}) ${digits.slice(3, 6)}-${digits.slice(6)}`;
}
